tests/test_export.py

```diff
@@ -412,3 +417,155 @@ def test_3787_key_error_export(snippetcompiler):
+        install_v2_modules=[LocalPackagePath(module_path, editable=False)],
+        project_requires=[InmantaModuleRequirement.parse("minimalv2module"), InmantaModuleRequirement.parse("std")],
+    )
+
+    await snippetcompiler_clean.do_export_and_deploy()
+    resources = await Resource.get_list(environment=environment)
+    assert len(resources) == 4
+    assert resources[0].resource_id_value == "the_resource_a"
+    assert resources[0].resource_set == "resource_set_1"
+    assert resources[1].resource_id_value == "the_resource_b"
+    assert resources[1].resource_set == "resource_set_2"
+    assert resources[2].resource_id_value == "the_resource_c"
+    assert resources[2].resource_set == "resource_set_1"
+    assert resources[3].resource_id_value == "the_resource_d"
+    assert resources[3].resource_set is None
+
+
+@pytest.mark.slowtest
+async def test_resource_in_multiple_resource_sets(snippetcompiler_clean, modules_v2_dir: str, tmpdir, environment) -> None:
+    """
+    test that an error is raised if a resource is in multiple
+    resource_sets
+    """
+    init_cf = """
+entity Res extends std::Resource:
+    string name
+end
+
+implement Res using std::none
+
+a = Res(name="the_resource_a")
+std::ResourceSet(name="resource_set_1", resources=[a])
+std::ResourceSet(name="resource_set_2", resources=[a])
+"""
+    init_py = """
+from inmanta.resources import (
+    Resource,
+    resource,
+)
+@resource("minimalv2module::Res", agent="name", id_attribute="name")
+class Res(Resource):
+    fields = ("name",)
+"""
+    module_name: str = "minimalv2module"
+    module_path: str = str(tmpdir.join(module_name))
+
+    module_from_template(
+        os.path.join(modules_v2_dir, module_name),
+        module_path,
+        new_requirements=[InmantaModuleRequirement.parse("std")],
+        new_content_init_cf=init_cf,
+        new_content_init_py=init_py,
+    )
+    snippetcompiler_clean.setup_for_snippet(
+        """
+import minimalv2module
+        """,
+        install_v2_modules=[LocalPackagePath(module_path, editable=False)],
+        project_requires=[InmantaModuleRequirement.parse("minimalv2module"), InmantaModuleRequirement.parse("std")],
+    )
+    with pytest.raises(CompilerException) as e:
+        await snippetcompiler_clean.do_export_and_deploy()
+    assert (
+        e.value.format_trace()
+        == "resource 'minimalv2module::Res[the_resource_a,name=the_resource_a]' can not be part of multiple ResourceSets"
+    )
+
+
+@pytest.mark.slowtest
+async def test_resource_not_exported(snippetcompiler_clean, modules_v2_dir: str, tmpdir, environment, caplog) -> None:
+    """
+    test that an error is raised if a resource that is not exported is in a resource_set
+    """
+    init_cf = """
+std::ResourceSet(name="resource_set_1", resources=[std::Resource()])
+implement std::Resource using std::none
+"""
+    module_name: str = "minimalv2module"
+    module_path: str = str(tmpdir.join(module_name))
+
+    module_from_template(
```

Comment:
Why go through the effort of setting up a module, if this whole module can just go into the main file (i.e. `setup_for_snippet` )

I think module_from_template is very slow and I see no reason to use it here. 

I think, if you drop all the modules v2 stuff, this will no longer be a slow test?

Also, why use `snippetcompiler_clean` instead of `snippetcompiler`? This clean version is a lot slower as well.

(sorry I didn't see this the first time around). 


(this applies to all testcases)

Reply:
@wouterdb Do I understand correctly that only the part about `snippetcompiler_clean` applies to all the test cases as for the first two the whole module isn't contained into the main file ?